src/axom/primal/geometry/KnotVector.hpp

```diff
@@ -0,0 +1,501 @@
+// Copyright (c) 2017-2024, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/*!
+ * \file NURBSCurve.hpp
+ *
+ * \brief A NURBS curve primitive
+ */
+
+#ifndef AXOM_PRIMAL_KNOTVECTOR_HPP
+#define AXOM_PRIMAL_KNOTVECTOR_HPP
+
+#include "axom/core.hpp"
+#include "axom/slic.hpp"
+
+#include <ostream>
+#include "axom/fmt.hpp"
+
+namespace axom
+{
+namespace primal
+{
+// Forward declare the templated classes and operator functions
+template <typename T>
+class KnotVector;
+
+/*! \brief Overloaded output operator for Bezier Curves*/
+template <typename T>
+std::ostream& operator<<(std::ostream& os, const KnotVector<T>& kVec);
+
+/*!
+ * \class KnotVector
+ *
+ * \brief Represents the knot vector for a B-Spline/NURBS Curve/Surface
+ * \tparam T the coordinate type, e.g., double, float, etc.
+ * 
+ * Contains methods for knot vector manipulation shared between NURBS curves and surfaces.
+ * This class maintains the degree and requirements on a knot vector, namely
+ *  > Monotonicity (u_i <= u_i+1)
+ *  > Open/Clamped-ness (degree + 1 knots at beginning and end)
+ *  > Continuity (maximum internal knot multiplicity = degree)
+ */
+template <typename T>
+class KnotVector
+{
+public:
+  AXOM_STATIC_ASSERT_MSG(
+    std::is_arithmetic<T>::value,
+    "A knot vector must be defined using an arithmetic type");
+
+public:
+  /// \brief Default constructor
+  KnotVector() : m_deg(-1) { m_knots.resize(0); }
+
+  /*!
+   * \brief Constructor for a normalized, uniform knot vector
+   *
+   * \param [in] npts the number of control points
+   * \param [in] deg the degree
+   * \pre degree is greater than or equal to -1, and deg < npts.
+   */
+  KnotVector(axom::IndexType npts, int deg) { makeUniform(npts, deg); }
+
+  /// \brief Constructor from a user-supplied knot vector (C-style)
+  KnotVector(const T* knots, axom::IndexType nkts, int degree)
+  {
+    m_knots.resize(nkts);
+    for(int i = 0; i < nkts; ++i)
+    {
+      m_knots[i] = knots[i];
+    }
+
+    m_deg = degree;
+  }
```

Comment:
I'm open to changing it, as long as it's consistent across the different classes. I wonder if it would be worth a separate PR to handle this and the different array accessors, or just put them all here?